app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/bugreport/NetworkTypeCollector.kt

```diff
@@ -111,34 +118,60 @@ class NetworkTypeCollector @Inject constructor(
+        private var privateDnsCacheValue: String? = null
+        private var cachedSystemDns: Set<String>? = null
+
         override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
             super.onLinkPropertiesChanged(network, linkProperties)
 
+            val shouldRestart = AtomicBoolean(false)
+
             // check for Android Private DNS setting
-            val privateDns = if (context.isPrivateDnsActive()) {
-                logcat {
-                    """
-                        isPrivateDnsActive = ${context.isPrivateDnsActive()}, server = ${context.getPrivateDnsServerName()}
-                        (${runCatching { InetAddress.getAllByName(context.getPrivateDnsServerName()) }.getOrNull()?.map { it.hostAddress }})
-                    """.trimIndent()
-                }
-                true
-            } else {
-                logcat { "Private DNS disabled" }
-                false
+            val privateDnsServerName = context.getPrivateDnsServerName()
+            val systemDnsNames = getSystemActiveNetworkDefaultDns()
+            logcat {
+                """
+                    isPrivateDnsActive = ${context.isPrivateDnsActive()}, server = ${context.getPrivateDnsServerName()}
+                    (${runCatching { InetAddress.getAllByName(context.getPrivateDnsServerName()) }.getOrNull()?.map { it.hostAddress }})
+                """.trimIndent()
             }
 
             // Check if VPN reconfiguration is needed
-            if (appTpFeatureConfig.isEnabled(AppTpSetting.PrivateDnsSupport) && vpnPreferences.isPrivateDnsEnabled != privateDns) {
-                logcat { "Private DNS changed, reconfiguring VPN" }
+            if (isPrivateDnsSupportEnabled && privateDnsCacheValue != privateDnsServerName) {
+                logcat { "Private DNS changed, should reconfigure VPN" }
+                shouldRestart.set(true)
+            } else if (isInterceptDnsTrafficSupported && didSystemDnsChanged(cachedSystemDns, systemDnsNames)) {
+                logcat { "System DNS changed, should reconfigure VPN, old = $cachedSystemDns, new=$systemDnsNames" }
+                shouldRestart.set(true)
+            }
+
+            // update values cached values anyways
+            privateDnsCacheValue = privateDnsServerName
+            cachedSystemDns = systemDnsNames
+            if (shouldRestart.getAndSet(false)) {
+                logcat { "Reconfiguring VPN now" }
                 coroutineScope.launch {
-                    vpnPreferences.isPrivateDnsEnabled = privateDns
                     TrackerBlockingVpnService.restartVpnService(context)
                 }
             } else {
                 logcat { "Nothing change in Network config, skip reconfiguration" }
             }
         }
+
+        private fun didSystemDnsChanged(oldSystemDns: Set<String>?, newSystemDns: Set<String>): Boolean {
+            if (oldSystemDns == null) return false
```

Comment:
`oldSystemDns` is `null` only when the VPN first starts. No need to reconfigure then